apps/emqx_mq/src/emqx_mq_consumer/emqx_mq_consumer_db.erl

```diff
@@ -0,0 +1,304 @@
+        fun(Shard) ->
+            Topic = ['#'],
+            TxOpts = #{
+                db => ?MQ_CONSUMER_DB,
+                shard => Shard,
+                generation => 1,
+                sync => true,
+                retries => 5
+            },
+            emqx_ds:trans(TxOpts, fun() ->
+                emqx_ds:tx_del_topic(Topic)
+            end)
+        end,
+        Shards
+    ).
+
+%%--------------------------------------------------------------------
+%% Internal functions
+%%--------------------------------------------------------------------
+
+init_consumer_data() ->
+    #{
+        progress => #{}
+    }.
+
+read_claim_tx(LeadershipTopic) ->
+    case emqx_ds:tx_read(LeadershipTopic) of
+        [] ->
+            not_found;
+        [{_Topic, 0, ClaimBin}] ->
+            {ok, decode_claim(ClaimBin)}
+    end.
+
+write_claim(LeadershipTopic, Claim) ->
+    emqx_ds:tx_write({LeadershipTopic, 0, encode_claim(Claim)}).
+
+delete_claim(LeadershipTopic) ->
+    emqx_ds:tx_del_topic(LeadershipTopic).
+
+read_or_init_consumer_data(DataTopic) ->
+    Res = emqx_ds:tx_read(DataTopic),
+    ?tp_debug(mq_consumer_db_read_or_init_consumer_data, #{data_topic => DataTopic, res => Res}),
+    case Res of
+        [] ->
+            InitData = init_consumer_data(),
+            ok = write_consumer_data(DataTopic, InitData),
+            InitData;
+        [{_Topic, 0, DataBin}] ->
+            decode_consumer_data(DataBin)
+    end.
+
+write_consumer_data(DataTopic, Data) ->
+    _TTV = {DataTopic, 0, Data},
+    ?tp_debug(mq_consumer_db_write_consumer_data, #{ttv => _TTV}),
+    emqx_ds:tx_write({DataTopic, 0, encode_consumer_data(Data)}).
+
+decode_claim(ClaimBin) ->
+    emqx_mq_consumer_db_serializer:decode_claim(ClaimBin).
+
+encode_claim(Claim) ->
+    emqx_mq_consumer_db_serializer:encode_claim(Claim).
+
+decode_consumer_data(DataBin) ->
+    emqx_mq_consumer_db_serializer:decode_consumer_data(DataBin).
+
+encode_consumer_data(Data) ->
+    emqx_mq_consumer_db_serializer:encode_consumer_data(Data).
+
+topic_leadership(#{topic_filter := MQTopic, id := MQId}) ->
+    ?TOPIC_LEADERSHIP(MQTopic, MQId).
+
+topic_consumer_data(#{topic_filter := MQTopic, id := MQId}) ->
+    ?TOPIC_CONSUMER_DATA(MQTopic, MQId).
+
+settings() ->
+    NSites = length(emqx:running_nodes()),
+    #{
+        transaction =>
+            #{
+                flush_interval => 10,
+                idle_flush_interval => 5,
+                conflict_window => 5000
+            },
+        storage =>
+            {emqx_ds_storage_skipstream_lts_v2, ?MQ_CONSUMER_DB_LTS_SETTINGS},
+        store_ttv => true,
+        backend => builtin_raft,
+        n_shards => NSites * ?SHARDS_PER_SITE,
+        replication_options => #{},
+        n_sites => NSites,
+        replication_factor => ?REPLICATION_FACTOR
+    }.
+
+tx_opts(#{topic_filter := MQTopic, id := MQId}) ->
+    #{
+        db => ?MQ_CONSUMER_DB,
+        shard => {auto, {MQTopic, MQId}},
```

Comment:
I'm not sure whether this API will always accept arbitrary terms instead of being limited to just binaries. The current relaxed behavior is sort of accidental. 

If you want to shard by a tuple, you're now free to implement the sharding strategy by yourself, this way you don't have to rely on 3rd party API.